Mathlib/RingTheory/Kaehler/CotangentComplex.lean

```diff
@@ -0,0 +1,276 @@
+
+## Main results
+- `Algebra.Generators.Cotangent`: The conormal space `I/I²`. (Defined in `Generators/Basic`)
+- `Algebra.Generators.CotangentSpace`: The cotangent space `⨁ᵢ S dxᵢ`.
+- `Algebra.Generators.CotangentComplex`: The map `I/I² → ⨁ᵢ S dxᵢ`.
+- `Algebra.Generators.toKaehler`: The projection `⨁ᵢ S dxᵢ → Ω[S/R]`.
+- `Algebra.Generators.toKaehler_surjective`: The map `⨁ᵢ S dxᵢ → Ω[S/R]` is surjective.
+- `Algebra.Generators.exact_cotangentComplex_toKaehler`: `I/I² → ⨁ᵢ S dxᵢ → Ω[S/R]` is exact.
+- `Algebra.Generators.Hom.Sub`: If `f` and `g` are two maps between presentations, `f - g` induces
+  a map `⨁ᵢ S dxᵢ → I/I²` that makes `f` and `g` homotopic.
+-/
+
+open KaehlerDifferential TensorProduct MvPolynomial
+
+namespace Algebra
+
+universe w u v
+
+variable {R : Type u} {S : Type v} [CommRing R] [CommRing S] [Algebra R S]
+
+namespace Generators
+
+variable (P : Generators.{w} R S)
+
+/--
+The cotangent space on `P = R[X]`.
+This is isomorphic to `Sⁿ` with `n` being the number of variables of `P`.
+-/
+abbrev CotangentSpace : Type _ := S ⊗[P.Ring] Ω[P.Ring⁄R]
+
+/-- The canonical basis on the `CotangentSpace`. -/
+noncomputable
+def cotangentSpaceBasis : Basis P.vars S P.CotangentSpace :=
+  (mvPolynomialBasis _ _).baseChange _
+
+@[simp]
+lemma cotangentSpaceBasis_repr_tmul (r x i) :
+    P.cotangentSpaceBasis.repr (r ⊗ₜ .D _ _ x) i = r * aeval P.val (pderiv i x) := by
+  classical
+  simp only [cotangentSpaceBasis, Basis.baseChange_repr_tmul, mvPolynomialBasis_repr_apply,
+    Algebra.smul_def, mul_comm r]
+  rfl
+
+lemma cotangentSpaceBasis_repr_one_tmul (x i) :
+    P.cotangentSpaceBasis.repr (1 ⊗ₜ .D _ _ x) i = aeval P.val (pderiv i x) := by
+  rw [cotangentSpaceBasis_repr_tmul, one_mul]
+
+lemma cotangentSpaceBasis_apply i :
+    P.cotangentSpaceBasis i = 1 ⊗ₜ .D _ _ (.X i) := by
+  simp [cotangentSpaceBasis]
+
+/-- The cotangent complex given by a presentation `R[X] → S` (i.e. a closed embedding `S ↪ Aⁿ`). -/
+noncomputable
+def cotangentComplex : P.Cotangent →ₗ[S] P.CotangentSpace :=
+  letI f : P.Cotangent ≃ₗ[P.Ring] P.ker.Cotangent :=
+    { __ := AddEquiv.refl _, map_smul' := Cotangent.val_smul' }
+  (kerCotangentToTensor R P.Ring S ∘ₗ f).extendScalarsOfSurjective P.algebraMap_surjective
+
+@[simp]
+lemma cotangentComplex_mk (x) : P.cotangentComplex (.mk x) = 1 ⊗ₜ .D _ _ x :=
+  kerCotangentToTensor_toCotangent _ _ _ _
+
+universe w' u' v'
+
+variable {R' : Type u'} {S' : Type v'} [CommRing R'] [CommRing S'] [Algebra R' S']
+variable (P' : Generators.{w'} R' S')
+variable [Algebra R R'] [Algebra S S'] [Algebra R S'] [IsScalarTower R R' S'] [IsScalarTower R S S']
+
+attribute [local instance] SMulCommClass.of_commMonoid
+
+variable {P P'}
+
+/--
+This is the map on the cotangent space associated to a map of presentation.
+The matrix associated to this map is the Jacobian matrix. See `CotangentSpace.repr_map`.
+-/
+noncomputable
+def CotangentSpace.map (f : Hom P P') : P.CotangentSpace →ₗ[S] P'.CotangentSpace := by
```

Comment:
Would you mind starting a namespace, to shorten a whole bunch of names in this file?